fix latent heat viscosity prefactor

jdannberg opened the pull request:
The list of viscosity prefactors that would change the viscosity at phase transitions in the latent heat material model did not work correctly if more than one compositional field was used. Specifically, individual layers would have the wrong viscosities, and the length of the list of prefactors that had to be specified in the input file made no sense. This specific setup had no test case, which is likely why we did not notice when we broke this. 
The problem was that the prefactors are specified relative to each other, but when they are applied, they are multiplied on top of each other (due to the way the phase function works). This means they have to be scaled after they are read in, and this scaling depends on how many phase transitions are occurring above a given phase, and the old interface just had no good way of figuring that out. Therefore, there was no good way for fixing this without also changing the interface for this input parameter. I changed the interface to be consistent with what we use for all other phase transition parameters (the map parsing from the material model utilities), which has the advantage that there's a good default for models with and without chemical heterogeneities. I also fixed the computation of the viscosities in the individual phases, and changed the averaging between phases to a log average (which is consistent with the rest of the material model, which already uses log averaging for viscosities). 

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

tjhei commented on source/material_model/latent_heat.cc:
```diff
@@ -131,18 +136,27 @@ namespace aspect
+
+                // Viscosities
+                if (transition_index+1 < phase_function.n_phases_for_each_composition()[0])      // 1st compositional field
+                  viscosity_phase_prefactors[0] = std::pow(10,
```
```suggestion
                  viscosity_phase_prefactors[0] = std::pow(10.,
```

tjhei commented on source/material_model/latent_heat.cc:
```diff
@@ -131,18 +136,27 @@ namespace aspect
+                                                           + (1. - phaseFunction) * std::log10(viscosity_phase_prefactors[0]));
+                else if (transition_index+2 < phase_function.n_phases_for_each_composition()[0]
+                         + phase_function.n_phases_for_each_composition()[1])                  // 2nd compositional field
+                  viscosity_phase_prefactors[1] = std::pow(10,
```
```suggestion
                  viscosity_phase_prefactors[1] = std::pow(10.,
```

tjhei commented on source/material_model/latent_heat.cc:
```diff
@@ -164,7 +178,12 @@ namespace aspect
+            if (composition.size()==0)
+              out.viscosities[i] = out.viscosities[i]*viscosity_phase_prefactors[0];
+            else
+              out.viscosities[i] = (std::pow(10, ((1-composition[0]) * std::log10(out.viscosities[i]*viscosity_phase_prefactors[0])
```
```suggestion
              out.viscosities[i] = (std::pow(10., ((1-composition[0]) * std::log10(out.viscosities[i]*viscosity_phase_prefactors[0])
```

tjhei commented on source/material_model/latent_heat.cc:
```diff
@@ -302,12 +321,14 @@ namespace aspect
+                             "viscosity (modified by any compositional prefactors) will be multiplied "
+                             "by this factor to get the corresponding viscosity for each phase. "
+                             "List must have the same number of entries as there are phases, that is one "
+                             "more than Phase transition depths for each composition that is used in the "
```
do you want to quote this parameter?

jdannberg commented:
@tjhei Thank you for the review!
I've addressed your comments and updated the one test that was failing. I checked that the results are still correct (the small change likely comes from fixing the averaging across phase transitions to use a log average rather than an arithmetic average).

gassmoeller reviewed (approved):
Yes, clearly a fix, even though I find the new code hard to understand as well.
